joomla/joomla-cms
# Add attribs to the script tag, add version to JHtml script methods and deprecate MD5SUM (alternative to 8540)

### Summary of Changes

This PR (alternative to https://github.com/joomla/joomla-cms/pull/8540) adds the ability to add more attributes to the script tag using `JDocument::addScript()`, `JDocument::addScriptVersion()` and `JHtml::script()` methods.  Also adds the ability to add IE conditional statements.

Also adds the ability to add the same attribs, script version and IE conditional statements when using `JHtml::script()` method.

Also deprecates MD5SUM and JDocument::addScriptVersion method.
### How to test

Please test this in all possible scenarios.
Any problem will get scripts not loading.
- Use latest staging
- Apply patch
- Check backend and frontend are working fine
- Add external js scripts like to isis/protostar index.php, some examples:

``` php
// JDocument B/C tests
$this->addScript('/path/to/external/bcaddScript1.js');
$this->addScript('/path/to/external/bcaddScript2.js', '', true, false);
$this->addScript('/path/to/external/bcaddScript3.js', 'text/javascript', true, true);
$this->addScriptVersion('/path/to/external/bcaddScriptVersion1.js');
$this->addScriptVersion('/path/to/external/bcaddScriptVersion2.js', null, '', true, false);
$this->addScriptVersion('/path/to/external/bcaddScriptVersion3.js', 'version-string', 'text/javascript', true, true);

// JDocument new method tests
$this->addScript('/path/to/external/addScript0.js');
$this->addScript('/path/to/external/addScript1.js', array(), array('id' => 'scriptid'));
$this->addScript('/path/to/external/addScript2.js', array(), array('id' => 'scriptid', 'data-test' => 1));
$this->addScript('/path/to/external/addScript3.js', array(), array('data-attrib' => 'original-value'));
$this->addScript('/path/to/external/addScript4.js', array(), array('data-attrib' => 'final-value'));
$this->addScript('/path/to/external/addScript5.js', array(), array('data-attrib' => array('item1' => 'value1', 'item2' => 'value2')));
$this->addScript('/path/to/external/addScript6.js', array('version' => 'auto'), array('data-attrib' => array('item1' => 'value1', 'item2' => 'value2')));
$this->addScript('/path/to/external/addScript7.js', array('version' => 'version-string'), array('data-attrib' => array('item1' => 'value1', 'item2' => 'value2')));
$this->addScript('/path/to/external/addScript8.js', array('version' => 'auto', 'conditional' => 'lf IE 9'), array('data-attrib' => array('item1' => 'value1', 'item2' => 'value2')));

// JHTML B/C Tests
JHtml::_('script', 'media/system/js/multiselect.js');
JHtml::_('script', 'system/repeatable.js', false, true, false, false, true);

// JHtml new method tests
JHtml::_('script', 'media/com_finder/js/indexer.js', array(), array('data-attrib' => array('item1' => 3, 'item2' => 'value2')));
JHtml::_('script', 'system/progressbar.js', array('relative' => true), array('data-attrib' => array('item1' => 3, 'item2' => 'value2')));
JHtml::_('script', 'system/permissions.js', array('relative' => true, 'version' => 'auto'), array('data-attrib' => array('item1' => 3, 'item2' => 'value2')));
JHtml::_('script', 'system/sendtestmail.js', array('relative' => true, 'version' => 'version-string'), array('async' => 'async', 'defer' => 'defer', 'data-attrib' => array('item1' => 3, 'item2' => 'value2')));
JHtml::_('script', 'system/helpsite.js', array('relative' => true, 'version' => 'version-string', 'conditional' => 'lf IE 9'), array('async' => 'async', 'defer' => 'defer', 'data-attrib' => array('item1' => 3, 'item2' => 'value2')));
```
- Check the HTML output and confirm all is fine
- do the same tests for anything else you can remember
- Do a code review.
### Observations

From a B/C perspective i think there are no issues, but mantainers please check.
### PRs depending on this one getting tested & merged
- https://github.com/joomla/joomla-cms/pull/11686
- https://github.com/joomla/joomla-cms/pull/8545
